arrow-select/src/coalesce.rs

```diff
@@ -0,0 +1,596 @@
+    /// clears any output buffers
+    ///
+    /// Normally this is called when the input stream is exhausted, and
+    /// we want to finalize the last batch of rows.
+    ///
+    /// See [`Self::next_batch()`] for the completed batches.
+    pub fn finish_batch(&mut self) -> Result<(), ArrowError> {
+        let batch = concat_batches(&self.schema, &self.buffer)?;
+        self.buffer.clear();
+        self.buffered_rows = 0;
+        self.completed.push_back(batch);
+        Ok(())
+    }
+
+    /// Returns the next completed batch, if any
+    pub fn next_batch(&mut self) -> Option<RecordBatch> {
+        self.completed.pop_front()
+    }
+}
+
+/// Heuristically compact `StringViewArray`s to reduce memory usage, if needed
+///
+/// Decides when to consolidate the StringView into a new buffer to reduce
+/// memory usage and improve string locality for better performance.
+///
+/// This differs from `StringViewArray::gc` because:
+/// 1. It may not compact the array depending on a heuristic.
+/// 2. It uses a precise block size to reduce the number of buffers to track.
+///
+/// # Heuristic
+///
+/// If the average size of each view is larger than 32 bytes, we compact the array.
+///
+/// `StringViewArray` include pointers to buffer that hold the underlying data.
+/// One of the great benefits of `StringViewArray` is that many operations
+/// (e.g., `filter`) can be done without copying the underlying data.
+///
+/// However, after a while (e.g., after `FilterExec` or `HashJoinExec`) the
+/// `StringViewArray` may only refer to a small portion of the buffer,
+/// significantly increasing memory usage.
+fn gc_string_view_batch(batch: &RecordBatch) -> RecordBatch {
+    let new_columns: Vec<ArrayRef> = batch
+        .columns()
+        .iter()
+        .map(|c| {
+            // Try to re-create the `StringViewArray` to prevent holding the underlying buffer too long.
+            let Some(s) = c.as_string_view_opt() else {
+                return Arc::clone(c);
+            };
+            let ideal_buffer_size: usize = s
+                .views()
+                .iter()
+                .map(|v| {
+                    let len = (*v as u32) as usize;
+                    if len > 12 {
+                        len
+                    } else {
+                        0
+                    }
+                })
+                .sum();
+            let actual_buffer_size = s.get_buffer_memory_size();
+
+            // Re-creating the array copies data and can be time consuming.
+            // We only do it if the array is sparse
+            if actual_buffer_size > (ideal_buffer_size * 2) {
+                // We set the block size to `ideal_buffer_size` so that the new StringViewArray only has one buffer, which accelerate later concat_batches.
+                // See https://github.com/apache/arrow-rs/issues/6094 for more details.
+                let mut builder = StringViewBuilder::with_capacity(s.len());
```

Comment:
Reusing the views is likely quite a bit faster (didn't test the transmute, I think that should go to from / as_u128 if it shows to be faster in benchmarks):
```suggestion
                let mut buffer: Vec<u8> = Vec::with_capacity(ideal_buffer_size);

                let views: Vec<u128> = s.views().as_ref().iter().cloned().map(|v| {
                    // SAFETY: ByteView has same memory layout as u128
                    let mut b: ByteView = unsafe { std::mem::transmute(v) };

                    if b.length > 12 {
                        let offset = buffer.len() as u32;
                        buffer.extend_from_slice(
                            buffers[b.buffer_index as usize]
                                .get(b.offset as usize..b.offset as usize + b.length as usize)
                                .expect("Invalid buffer slice"),
                        );
                        b.offset = offset;
                        b.buffer_index = 0; // Set buffer index to 0, as we only have one buffer
                    }

                    unsafe { std::mem::transmute(b) }
                    
                }).collect();```